src/routes/safe/components/Apps/components/AppFrame.tsx

```diff
@@ -170,6 +171,14 @@ const AppFrame = ({ appUrl }: Props): ReactElement => {
       chainId: NETWORK_ID,
     }))
 
+    communicator?.on('getSafeBalances', async (msg) => {
+      const { currency = 'usd' } = msg.data.params as GetBalanceParams
+
+      const balances = await fetchTokenCurrenciesBalances({ safeAddress, selectedCurrency: currency })
```

Comment:
Can we not reuse the local state of the balances? We regularly update the state and it would be nice to make use of that instead of triggering another request.

Also I would add support for pagination, even if we don't use it right now. But might also be enough to document this on sdk side. Devs should not expect that all balances are available with one request.

Reply:
> Can we not reuse the local state of the balances?

I will check it, but I'm not sure if on the apps page we perform the updates

Reply:
Also, the way it's implemented right now (useEffect hook), usage of state piece (balances) would mean the communicator would mount/unmount on state changes, which could lead to some app's messages being lost. I wanted to rewrite it in a way so communicator would spawn only once but haven't got the chance to do that yet.

Also, cmon, it's just one request 😛

Reply:
If we could guarantee that it is only 1 request this would be no issue. We are currently working on lowering the requests the interface is making and this goes in a different direction. If an app wants to they can query this every second. This should not have a big impact, but it would be annoying. I actually would like some other measures against this: e.g. add the safe app url as a header field, this way we could track where teaffic is coming from if required

Reply:
>  If an app wants to they can query this every second. 

It seems to be related to the rate-limiting discussion we previously had. I think it's the thing to solve generally, not for this particular request

Reply:
> Also I would add support for pagination

I checked the gateway code and it doesn't seem to have pagination for this request. Also, safe-react doesn't use it

Reply:
> rate limiting

If you use a cached state then it doesn't matter. In general it is less critical here than for the rpc calls

> No pagination support

That is currently true, but might not always be the case. If we change this it would mean that all existing Safe apps need to be updated or they might not get complete data.

Reply:
> That is currently true, but might not always be the case. If we change this it would mean that all existing Safe apps need to be updated or they might not get complete data.

I'm not quite sure how we can add support for pagination at this point then. If we implement it on the backend, we can implement call handler on the interface in a way that it would fetch all pages and return them to the app

Reply:
> If we implement it on the backend, we can implement a call handler on the interface in a way that it would fetch all pages and return them to the app

That defeats the purpose of pagination. We always implement pagination with the same scheme: you haven an array of `results` and a `next` url that allows you to fetch more. But as mentioned before it is also fine not to implement, but then we should document, that we only return the top tokens for this account (whatever this might mean in the future).

I feel like we jumped into the implementation of this feature without talking about the details. We are creating a strong dependency to our balances endpoint, and we should be aware of this.

Reply:
> That defeats the purpose of pagination.

True, but I have never seen a paginated token list in any dapp